File: app/src/main/java/org/eztarget/papeler/BackgroundImageOpener.java
package org.eztarget.papeler;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.RectF;
import android.os.Environment;
import android.util.Log;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.InputStream;

/**
 * Created by michelsievers on 14/03/2017.
 */

class BackgroundImageOpener {

    private static final String TAG = BackgroundImageOpener.class.getSimpleName();

    static Bitmap load() throws Exception {

        final File imageFile = getFile();
        if (imageFile == null) {
            Log.e(TAG, "Image file is null.");
            throw new FileNotFoundException("Image File is null");
        }

        final InputStream imageStream = new FileInputStream(getFile());
        return BitmapFactory.decodeStream(imageStream);
    }

    static Bitmap loadRatioPreserved(final int targetWidth, final int targetHeight)
            throws Exception {

        final Bitmap bitmap = load();
        final int sourceWidth = bitmap.getWidth();
        final int sourceHeight = bitmap.getHeight();

        final float xScale = (float) targetWidth / sourceWidth;
        final float yScale = (float) targetHeight / sourceHeight;
        final float scale = Math.max(xScale, yScale);

        final float scaledWidth = scale * sourceWidth;
        final float scaledHeight = scale * sourceHeight;

        final float left = (targetWidth - scaledWidth) / 2;
        final float top = (targetHeight - scaledHeight) / 2;

        final RectF targetRect = new RectF(left, top, left + scaledWidth, top + scaledHeight);

        final Bitmap dest = Bitmap.createBitmap(targetWidth, targetHeight, bitmap.getConfig());
        final Canvas canvas = new Canvas(dest);
        canvas.drawBitmap(bitmap, null, targetRect, null);

        return dest;
    }

    static File getFile() {
        final File destinationDir;
        destinationDir = new File(Environment.getExternalStorageDirectory().getPath() + "/waypr/");
        if (destinationDir.exists()) {
            Log.d(TAG, "Directory " + destinationDir.getAbsolutePath() + " exists.");
        } else {
            Log.d(TAG, "Creating " + destinationDir.getAbsolutePath() + ".");
            if (!destinationDir.mkdirs()) {
                Log.e(TAG, "Could not create " + destinationDir.getAbsolutePath() + ".");
                return null;
            }
        }

        return new File(destinationDir, "w.png");
    }

}

File: app/src/main/java/org/eztarget/papeler/BackgroundImageOpener.kt
package org.eztarget.papeler

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.RectF
import android.os.Environment
import android.util.Log

import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.InputStream

internal class BackgroundImageOpener {

    val storageDirRootPath: String = Environment.getExternalStorageDirectory().path
    val storageFileName = STORAGE_FILE_NAME

    val file: File?
        get() {
            val storageDirPath = storageDirRootPath + STORAGE_DIR_RELATIVE_PATH
            val destinationDir = File(storageDirPath)
            if (destinationDir.exists()) {
                Log.d(TAG, "Directory " + destinationDir.absolutePath + " exists.")
            } else {
                Log.d(TAG, "Creating " + destinationDir.absolutePath + ".")
                if (!destinationDir.mkdirs()) {
                    Log.e(TAG, "Could not create " + destinationDir.absolutePath + ".")
                    return null
                }
            }

            return File(destinationDir, storageFileName)
        }

    @Throws(Exception::class)
    fun load(): Bitmap {

        val imageFile = file
        if (imageFile == null) {
            Log.e(TAG, "Image file is null.")
            throw FileNotFoundException("Image File is null")
        }

        val imageStream = FileInputStream(file)
        return BitmapFactory.decodeStream(imageStream)
    }

    @Throws(Exception::class)
    fun loadRatioPreserved(targetWidth: Int, targetHeight: Int): Bitmap {
        val bitmap = load()
        val sourceWidth = bitmap.width
        val sourceHeight = bitmap.height

        val xScale = targetWidth.toFloat() / sourceWidth
        val yScale = targetHeight.toFloat() / sourceHeight
        val scale = Math.max(xScale, yScale)

        val scaledWidth = scale * sourceWidth
        val scaledHeight = scale * sourceHeight

        val left = (targetWidth - scaledWidth) / 2
        val top = (targetHeight - scaledHeight) / 2

        val targetRect = RectF(left, top, left + scaledWidth, top + scaledHeight)

        val dest = Bitmap.createBitmap(targetWidth, targetHeight, bitmap.config)
        val canvas = Canvas(dest)
        canvas.drawBitmap(bitmap, null, targetRect, null)

        return dest
    }

    companion object {
        private val TAG = BackgroundImageOpener::class.java.simpleName
        private const val STORAGE_DIR_RELATIVE_PATH = "/waypr/"
        private const val STORAGE_FILE_NAME = "w.png"
    }

}
